src/sage/modules/with_basis/subquotient.py

```diff
@@ -377,15 +377,217 @@ def is_submodule(self, other):
             sage: H.is_submodule(F)
             False
         """
-        if other is self.ambient():
+        if other is self._ambient:
             return True
         if not isinstance(self, SubmoduleWithBasis) and self.ambient() is other.ambient():
             raise ValueError("other (=%s) should be a submodule of the same ambient space" % other)
         if self not in ModulesWithBasis.FiniteDimensional:
-            raise NotImplementedError("is_submodule for infinite dimensional modules")
+            raise NotImplementedError("only implemented for finite dimensional submodules")
+        if self.dimension() > other.dimension():  # quick dimension check
+            return False
         for b in self.basis():
             try:
                 other.retract(b.lift())
             except ValueError:
                 return False
         return True
+
+    def _common_submodules(self, other):
+        """
+        Helper method to return a pair of submodules of the same ambient
+        free modules to do the corresponding linear algebra.
+        """
+        from sage.modules.free_module import FreeModule
+        supp_order = self._support_order
+        A = FreeModule(self.base_ring(), len(supp_order))
+        U = A.submodule([A([vec[supp] for supp in supp_order]) for vec in self._basis], check=False)
+        V = A.submodule([A([vec[supp] for supp in supp_order]) for vec in other._basis], check=False)
+        return (U, V)
+
+    def is_equal_subspace(self, other):
+        r"""
+        Return whether ``self`` is an equal submodule to ``other``.
+
+        .. NOTE::
+
+            This is the mathematical notion of equality (as sets that are
+            isomorphic as vector spaces), which is weaker than the `==`
+            which takes into account things like the support order.
```

Comment:
I must admit that I fail to understand the notion of equality.  If I understand correctly, instances of `SubmoduleWithBasis` are equal if and only if they are identical (as `CombinatorialFreeModule`).  This means, that they have the same basis, the same prefix, and even the same LaTeX options:
```
            sage: F = CombinatorialFreeModule(QQ, ['a','b','c'])
            sage: G = CombinatorialFreeModule(QQ, ('a','b','c'))
            sage: F is G
            True

            sage: F = CombinatorialFreeModule(QQ, ['a','b','c'], latex_bracket=['LEFT', 'RIGHT'])
            sage: F.print_options()['latex_bracket']
            ('LEFT', 'RIGHT')

            sage: F is G
            False
```
(I think that the latter is *extremely* strange!)

I would have expected that 'mathematical equality' means that they have the same elements, i.e., their bases span the same space.  Am I mistaken?  The reason for my doubts is in lines 463 and 464:
```
if set(self._support_order) != set(other._support_order):  # different supports
    return False
```
Please clarify!

Reply:
> I must admit that I fail to understand the notion of equality. If I understand correctly, instances of `SubmoduleWithBasis` are equal if and only if they are identical (as `CombinatorialFreeModule`). This means, that they have the same basis, the same prefix, and even the same LaTeX options:

Yes, this is correct. We want parents to be unique; equal-but-not-identitical parents can be a pain to work with. For the most part, a user who is passing such options is likely meaning for them to be different. Yes, mathematically it is slightly strange, but I think it is quite natural from the programming side.

> I would have expected that 'mathematical equality' means that they have the same elements, i.e., their bases span the same space. Am I mistaken? The reason for my doubts is in lines 463 and 464:
> 
> ```
> if set(self._support_order) != set(other._support_order):  # different supports
>     return False
> ```
> 
> Please clarify!

I don't understand why you doubt this. This is a very quick check that the support of the subspace (the union of the supports of all of the vectors in the subspace) are equal. This is clearly a necessary condition for the subspaces to be equal, otherwise there is a vector in one subspace that uses a basis vector in the ambient space that never appears in the other.

Reply:
Ah, now it makes sense, thank you. I somehow thought that an element in self._support would be a list of coefficient, basis vector pairs. Sorry for the noise.